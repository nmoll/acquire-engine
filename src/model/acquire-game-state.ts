export type AcquireGameState = "not started" | "started";
