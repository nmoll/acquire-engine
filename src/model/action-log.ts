import { IGameState } from "./game-state";
import { PlayerAction } from "./player-action";
import { PlayerActionResult } from "./player-action-result";

export interface ActionLog {
  action: PlayerAction;
  actionResult: PlayerActionResult;
  state: IGameState;
}
