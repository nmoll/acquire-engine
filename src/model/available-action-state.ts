import { AvailableAction } from "./available-action";

export type IAvailableActionState = AvailableAction[];
