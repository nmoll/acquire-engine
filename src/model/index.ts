export * from "./board-square-state";
export * from "./game-state";
export * from "./hotel-chain-type";
export * from "./player-turn";
export * from "./shares-state";
