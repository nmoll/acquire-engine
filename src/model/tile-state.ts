export interface ITileState {
  availableTiles: number[];
  [playerId: number]: number[];
}
