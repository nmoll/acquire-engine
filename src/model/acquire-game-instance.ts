import { AcquireGameState } from "./acquire-game-state";

export interface IAcquireGameInstance {
  id: string;
  randomSeed: number;
  playerIds: string[];
  hostId: string;
  state: AcquireGameState;
  isOpen?: boolean;
  createdDate?: number;
}
