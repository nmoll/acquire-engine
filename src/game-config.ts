import { HotelChainType } from "./model";

export const GameConfig = {
  board: {
    size: 108,
  },
  cash: {
    startingAmount: 6000,
  },
  tile: {
    rackSize: 5,
  },
  turn: {
    maxShares: 3,
  },
  hotel: {
    shares: 25,
    starterBonus: 1,
    basePrice: {
      WORLDWIDE: 200,
      LUXOR: 200,
      FESTIVAL: 300,
      IMPERIAL: 300,
      AMERICAN: 300,
      CONTINENTAL: 400,
      TOWER: 400,
    } as Record<HotelChainType, number>,
  },
};
