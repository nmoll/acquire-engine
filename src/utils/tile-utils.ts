import { GameConfig } from "../game-config";
import { PlayerAction } from "../model/player-action";
import { ITileState } from "../model/tile-state";
import { ArrayUtils } from "./array-utils";

const TILES = ArrayUtils.makeNumArray(GameConfig.board.size);

const getSortedBag = (seed: number): number[] =>
  ArrayUtils.shuffle(TILES, seed);

const concatPlayerTiles = (tileState: ITileState): number[] =>
  Object.values(tileState).reduce((res, tiles) => res.concat(tiles), []);

const getMaxTileIdx = (tileBag: number[], tiles: number[]): number =>
  Math.max(-1, ...tiles.map((tile) => tileBag.indexOf(tile)));

const getNextTile = (
  tileBag: number[],
  tileState: ITileState,
  actions: PlayerAction[]
): number | undefined => {
  const tilesPlaced = actions.filter(
    (action) => action.type === "PlaceTile"
  ).length;
  const tilesHeld = concatPlayerTiles(tileState).length;
  if (tilesPlaced + tilesHeld >= tileBag.length) {
    return undefined;
  }

  return tileBag[getMaxTileIdx(tileBag, concatPlayerTiles(tileState)) + 1];
};

export const TileUtils = {
  getSortedBag,
  getNextTile,
};
