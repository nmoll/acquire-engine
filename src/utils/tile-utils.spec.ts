import { GameConfig } from "../game-config";
import { ITileState } from "../model/tile-state";
import { TileUtils } from "./tile-utils";

describe("TileUtils", () => {
  describe("getSortedBag", () => {
    it("should return a tile bag sorted by the given seed", () => {
      const tileBag = TileUtils.getSortedBag(1);
      expect(tileBag).toMatchSnapshot();
      expect(tileBag.length).toEqual(GameConfig.board.size);
    });

    it("should return a different tile bag sorted by another seed", () => {
      const tileBag = TileUtils.getSortedBag(866);
      expect(tileBag).toMatchSnapshot();
      expect(tileBag.length).toEqual(GameConfig.board.size);
    });
  });

  describe("getNextTile", () => {
    it("should return first tile in bag if players have no tiles", () => {
      const tileBag = [3, 4, 7, 1];
      const tileState: ITileState = {};

      expect(TileUtils.getNextTile(tileBag, tileState, [])).toEqual(3);
    });

    it("should return the next tile not yet held by a player", () => {
      const tileBag = [3, 4, 7, 1, 8, 10];
      const tileState: ITileState = {
        1: [3, 4],
        2: [7],
      };

      expect(TileUtils.getNextTile(tileBag, tileState, [])).toEqual(1);
    });
  });
});
