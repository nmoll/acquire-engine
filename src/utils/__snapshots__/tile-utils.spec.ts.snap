// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileUtils getSortedBag should return a tile bag sorted by the given seed 1`] = `
Array [
  11,
  7,
  63,
  76,
  46,
  107,
  15,
  71,
  75,
  50,
  106,
  29,
  81,
  79,
  101,
  65,
  60,
  32,
  70,
  21,
  30,
  39,
  93,
  19,
  97,
  26,
  73,
  18,
  9,
  31,
  85,
  99,
  48,
  33,
  4,
  58,
  87,
  90,
  27,
  41,
  20,
  77,
  12,
  94,
  1,
  5,
  78,
  105,
  91,
  74,
  64,
  83,
  89,
  61,
  88,
  40,
  56,
  95,
  51,
  37,
  2,
  82,
  8,
  84,
  96,
  69,
  80,
  102,
  22,
  10,
  53,
  17,
  104,
  57,
  34,
  54,
  92,
  59,
  43,
  52,
  86,
  24,
  66,
  49,
  36,
  3,
  16,
  25,
  68,
  35,
  6,
  45,
  14,
  23,
  38,
  13,
  47,
  44,
  98,
  42,
  72,
  62,
  28,
  100,
  103,
  55,
  0,
  67,
]
`;
